system-addon/lib/TopStoriesFeed.jsm

```diff
@@ -272,6 +289,11 @@ this.TopStoriesFeed = class TopStoriesFeed {
       case at.NEW_TAB_REHYDRATED:
         this.maybeAddSpoc(action.meta.fromTarget);
         break;
+      case at.SECTION_UPDATE:
+        if (action.data.id === "highlights") {
+          this.dedupeFromHighlights();
```

Comment:
Updating ```this.stories``` in response to SECTION_UPDATE doesn't guarantee the UI is updated and the modified ```this.stories``` are actually used. We would have to fire a SECTION_UPDATE event for the topstories section to make sure, see https://github.com/mozilla/activity-stream/pull/3474/files#diff-13ca5ab376313a4fb771454ba25daa01R72.

We'd only need to fire this event if ```dedupefromHighlights``` is invoked as a response to SECTION_UPDATE. Not needed when invoked from fetchStories (there the even is fired anyway).



In general, I am not qui

Reply:
yep. makes sense!

Reply:
Ah, and in the former case we probably should only fire an SECTION_UPDATE if the stories actually changed after deduping from highlights.